### PR TITLE
Allow `ValueSerializerImpl` and `ValueDeserializerImpl` impls to be re-entrant

### PR DESCRIPTION
**Pull request by nathanwhit**
Currently, if you were to call (for instance) `write_value` as part of a custom `write_host_object` implementation, you would be forced to create aliased exclusive references to the serializer (which is instant UB).

The underlying v8 de/serializers are re-entrant, so for the most part we simply need to pass around shared references instead of exclusive references.

Also adds some bindings to some missing functions, and makes `ValueSerializer` (and deserializer) garbage collectable.

### CODE REVIEW

File: src/value_serializer.rs [hunk 1]
```diff
@@ -317,183 +346,224 @@ impl<'a, 's> ValueSerializerHeap<'a, 's> {
...
-      cxx_value_serializer: CxxValueSerializer {
-        _cxx_vtable: CxxVTable(std::ptr::null()),
-      },
-      cxx_value_serializer_delegate: CxxValueSerializerDelegate {
-        _cxx_vtable: CxxVTable(std::ptr::null()),
-      },
-      buffer_size: 0,
-      context: scope.get_current_context(),
-    });
+    let value_serializer_heap_ptr =
```
Comment (nathanwhit) [inline src/value_serializer.rs:500]: Previously this was violating aliasing rules because of the deref through the `Box`. To avoid this, use raw pointers for initialization.

File: src/value_deserializer.rs [hunk 2]
```diff
@@ -28,35 +31,36 @@ pub struct CxxValueDeserializerDelegate {
...
-  let scope =
-    &mut crate::scope::CallbackScope::new(value_deserializer_heap.context);
-  let value_deserializer_impl =
-    value_deserializer_heap.value_deserializer_impl.as_mut();
-  match value_deserializer_impl.read_host_object(
-    scope,
-    &mut value_deserializer_heap.cxx_value_deserializer,
-  ) {
+  let value_deserializer_heap = ValueDeserializerHeap::dispatch(this);
+  let scope = &mut CallbackScope::new(isolate.as_mut().unwrap());
```
Comment (devsnek) [inline src/value_deserializer.rs:38]: without the cppgc migration i think you can undo the changes here

**Review by devsnek — APPROVED** [review]

File: tests/test_api.rs [hunk 3]
```diff
@@ -7952,30 +7953,34 @@ fn bigint() {
 type ArrayBuffers = Vec<v8::SharedRef<v8::BackingStore>>;
 
 struct Custom1Value<'a> {
-  array_buffers: &'a mut ArrayBuffers,
+  array_buffers: RefCell<&'a mut ArrayBuffers>,
```
[the remark this answers is not in the document]
Comment (nathanwhit) [reply tests/test_api.rs:7955]: You actually can't mutate through an `&&mut` ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=0032e18dd880bb15fa0578716fdae403))

File: src/value_deserializer.rs (hunk 2 above)
Comment (nathanwhit) [reply src/value_deserializer.rs:38]: I think it's still needed, since you need to get a `Local<Context>` from a `Global<Context>`
